In [104]:
import mysql.connector
from mysql.connector import Error
import pymongo
import pandasql as ps
import pandas as pd
import numpy as np
from sqlalchemy import create_engine

## Using Pandasql library

In [84]:
# !pip install -U pandasql

In [85]:
sql_query = "select * from df_fitbit limit 3"

ps.sqldf(sql_query,locals())

Id                ActivityDate  TotalSteps  TotalDistance  \
0  1503960366  2016-03-25 00:00:00.000000       11004           7.11   
1  1503960366  2016-03-26 00:00:00.000000       17609          11.55   
2  1503960366  2016-03-27 00:00:00.000000       12736           8.53   

   TrackerDistance  LoggedActivitiesDistance  VeryActiveDistance  \
0             7.11                       0.0                2.57   
1            11.55                       0.0                6.92   
2             8.53                       0.0                4.66   

   ModeratelyActiveDistance  LightActiveDistance  SedentaryActiveDistance  \
0                      0.46                 4.07                      0.0   
1                      0.73                 3.91                      0.0   
2                      0.16                 3.71                      0.0   

   VeryActiveMinutes  FairlyActiveMinutes  LightlyActiveMinutes  \
0                 33                   12                   205   
1                 89                   17                   274   
2                 56                    5                   268   

   SedentaryMinutes  Calories  
0               804      1819  
1               588      2154  
2               605      1944

# Reading the data and performing basic operations to get an overview about the data set

In [72]:
df_fitbit = pd.read_csv('../../FSDS_DATA_SET/FitBit data.csv')

In [73]:
df_fitbit.shape

(457, 15)

In [74]:
df_fitbit.describe().transpose()

count          mean           std           min  \
Id                        457.0  4.628595e+09  2.293781e+09  1.503960e+09   
TotalSteps                457.0  6.546562e+03  5.398493e+03  0.000000e+00   
TotalDistance             457.0  4.663523e+00  4.082072e+00  0.000000e+00   
TrackerDistance           457.0  4.609847e+00  4.068540e+00  0.000000e+00   
LoggedActivitiesDistance  457.0  1.794274e-01  8.492318e-01  0.000000e+00   
VeryActiveDistance        457.0  1.180897e+00  2.487159e+00  0.000000e+00   
ModeratelyActiveDistance  457.0  4.786433e-01  8.309952e-01  0.000000e+00   
LightActiveDistance       457.0  2.890197e+00  2.237523e+00  0.000000e+00   
SedentaryActiveDistance   457.0  1.903720e-03  8.486801e-03  0.000000e+00   
VeryActiveMinutes         457.0  1.662363e+01  2.891970e+01  0.000000e+00   
FairlyActiveMinutes       457.0  1.307002e+01  3.620864e+01  0.000000e+00   
LightlyActiveMinutes      457.0  1.700700e+02  1.222054e+02  0.000000e+00   
SedentaryMinutes          457.0  9.952823e+02  3.370214e+02  3.200000e+01   
Calories                  457.0  2.189453e+03  8.154845e+02  0.000000e+00   

                                   25%           50%           75%  \
Id                        2.347168e+09  4.057193e+09  6.391747e+09   
TotalSteps                1.988000e+03  5.986000e+03  1.019800e+04   
TotalDistance             1.410000e+00  4.090000e+00  7.160000e+00   
TrackerDistance           1.280000e+00  4.090000e+00  7.110000e+00   
LoggedActivitiesDistance  0.000000e+00  0.000000e+00  0.000000e+00   
VeryActiveDistance        0.000000e+00  0.000000e+00  1.310000e+00   
ModeratelyActiveDistance  0.000000e+00  2.000000e-02  6.700000e-01   
LightActiveDistance       8.700000e-01  2.930000e+00  4.460000e+00   
SedentaryActiveDistance   0.000000e+00  0.000000e+00  0.000000e+00   
VeryActiveMinutes         0.000000e+00  0.000000e+00  2.500000e+01   
FairlyActiveMinutes       0.000000e+00  1.000000e+00  1.600000e+01   
LightlyActiveMinutes      6.400000e+01  1.810000e+02  2.570000e+02   
SedentaryMinutes          7.280000e+02  1.057000e+03  1.285000e+03   
Calories                  1.776000e+03  2.062000e+03  2.667000e+03   

                                   max  
Id                        8.877689e+09  
TotalSteps                2.849700e+04  
TotalDistance             2.753000e+01  
TrackerDistance           2.753000e+01  
LoggedActivitiesDistance  6.727057e+00  
VeryActiveDistance        2.192000e+01  
ModeratelyActiveDistance  6.400000e+00  
LightActiveDistance       1.251000e+01  
SedentaryActiveDistance   1.000000e-01  
VeryActiveMinutes         2.020000e+02  
FairlyActiveMinutes       6.600000e+02  
LightlyActiveMinutes      7.200000e+02  
SedentaryMinutes          1.440000e+03  
Calories                  4.562000e+03

In [75]:
df_fitbit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 457 entries, 0 to 456
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Id                        457 non-null    int64  
 1   ActivityDate              457 non-null    object 
 2   TotalSteps                457 non-null    int64  
 3   TotalDistance             457 non-null    float64
 4   TrackerDistance           457 non-null    float64
 5   LoggedActivitiesDistance  457 non-null    float64
 6   VeryActiveDistance        457 non-null    float64
 7   ModeratelyActiveDistance  457 non-null    float64
 8   LightActiveDistance       457 non-null    float64
 9   SedentaryActiveDistance   457 non-null    float64
 10  VeryActiveMinutes         457 non-null    int64  
 11  FairlyActiveMinutes       457 non-null    int64  
 12  LightlyActiveMinutes      457 non-null    int64  
 13  SedentaryMinutes          457 non-null    int64  
 14  Calories  

In [76]:
df_fitbit.head()

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366    3/25/2016       11004           7.11             7.11   
1  1503960366    3/26/2016       17609          11.55            11.55   
2  1503960366    3/27/2016       12736           8.53             8.53   
3  1503960366    3/28/2016       13231           8.93             8.93   
4  1503960366    3/29/2016       12041           7.85             7.85   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.57                      0.46   
1                       0.0                6.92                      0.73   
2                       0.0                4.66                      0.16   
3                       0.0                3.19                      0.79   
4                       0.0                2.16                      1.09   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 4.07                      0.0                 33   
1                 3.91                      0.0                 89   
2                 3.71                      0.0                 56   
3                 4.95                      0.0                 39   
4                 4.61                      0.0                 28   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   12                   205               804      1819  
1                   17                   274               588      2154  
2                    5                   268               605      1944  
3                   20                   224              1080      1932  
4                   28                   243               763      1886

In [77]:
df_fitbit['Id'].value_counts()

4057192912    32
4020332650    32
1503960366    19
1624580081    19
2347167796    15
4702921684    15
4445114986    15
6962181067    14
2320127002    12
4558609924    12
8792009665    12
8378563200    12
8253242879    12
7086361926    12
7007744171    12
2026352035    12
5553957443    12
8877689391    12
1844505072    12
2022484408    12
2873212765    12
3977333714    12
1927972279    12
4319703577    12
8053475328    11
5577150313    11
6290855005    10
1644430081    10
6117666160    10
3372868164    10
6391747486     9
6775888955     9
2891001357     8
8583815059     8
4388161847     8
Name: Id, dtype: int64

In [86]:
pd.set_option('display.max_columns', None)

df_fitbit.head()

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366   2016-03-25       11004           7.11             7.11   
1  1503960366   2016-03-26       17609          11.55            11.55   
2  1503960366   2016-03-27       12736           8.53             8.53   
3  1503960366   2016-03-28       13231           8.93             8.93   
4  1503960366   2016-03-29       12041           7.85             7.85   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.57                      0.46   
1                       0.0                6.92                      0.73   
2                       0.0                4.66                      0.16   
3                       0.0                3.19                      0.79   
4                       0.0                2.16                      1.09   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 4.07                      0.0                 33   
1                 3.91                      0.0                 89   
2                 3.71                      0.0                 56   
3                 4.95                      0.0                 39   
4                 4.61                      0.0                 28   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   12                   205               804      1819  
1                   17                   274               588      2154  
2                    5                   268               605      1944  
3                   20                   224              1080      1932  
4                   28                   243               763      1886

In [87]:
df_fitbit['ActivityDate'] = pd.to_datetime(df_fitbit['ActivityDate'], yearfirst=True)

### Working with mysql

In [102]:

def create_connection(host_name, user_name, password, database):

    try:
        connection = mysql.connector.connect(host=host_name, user = user_name, password=password, use_pure=True, auth_plugin='mysql_native_password')
    except Error as err:
        print('Error : ',err)
    else:
        return connection
        
def execute_query(connection, query:str,commit:bool=False):
    cursor = connection.cursor()
    result = None
    try:
        cursor.execute(query)
        result = cursor.fetchall()
        if commit:
            connection.commit()
        print("Query was successfully executed")
    except Error as err:
        print('Error : ',err)
    else:
        return result
    

def close_connection(connection):
    # cursor.close()
    connection.close()



        
        

In [103]:
host_name = 'localhost'
user_name = 'root'
password = '<password>'

conn = create_connection(host_name, user_name, password, 'dummytest')

print(conn)

In [101]:
query = 'select * from dummytest.ui_attribute'

# execute_query(conn,query)

### Inserting data into mysql 

> Database : ineuron_data, Table : fitbit

In [117]:
df_fitbit.head(2)

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366   2016-03-25       11004           7.11             7.11   
1  1503960366   2016-03-26       17609          11.55            11.55   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.57                      0.46   
1                       0.0                6.92                      0.73   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 4.07                      0.0                 33   
1                 3.91                      0.0                 89   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   12                   205               804      1819  
1                   17                   274               588      2154

In [109]:
engine = create_engine("mysql+pymysql://root:<password>@localhost:3306/ineuron_data")

try:
    df_fitbit.to_sql(name='fitbit', con=engine, index=False, if_exists='replace')
except Exception as e:
    print("Error is: ", e)
else:
    print("Query is Successful")

Query is Successful


In [116]:
close_connection(conn)

### Reading data from mysql

In [110]:
read_query = """
    SELECT * FROM ineuron_data.fitbit
"""

res = execute_query(conn,read_query)

Query was successfully executed


In [113]:
# for result in res:
#     print(result)

In [114]:
fitbit_sql = pd.read_sql(read_query, con=conn)

In [115]:
fitbit_sql.head()

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366   2016-03-25       11004           7.11             7.11   
1  1503960366   2016-03-26       17609          11.55            11.55   
2  1503960366   2016-03-27       12736           8.53             8.53   
3  1503960366   2016-03-28       13231           8.93             8.93   
4  1503960366   2016-03-29       12041           7.85             7.85   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.57                      0.46   
1                       0.0                6.92                      0.73   
2                       0.0                4.66                      0.16   
3                       0.0                3.19                      0.79   
4                       0.0                2.16                      1.09   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 4.07                      0.0                 33   
1                 3.91                      0.0                 89   
2                 3.71                      0.0                 56   
3                 4.95                      0.0                 39   
4                 4.61                      0.0                 28   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   12                   205               804      1819  
1                   17                   274               588      2154  
2                    5                   268               605      1944  
3                   20                   224              1080      1932  
4                   28                   243               763      1886

## Inserting data in Mongo Db

In [119]:
# df_fitbit.isna().sum()

`Converting 'df_fitbit' datafile into json format`

In [130]:
df_fitbit['ActivityDate'] = df_fitbit['ActivityDate'].dt.strftime('%Y-%m-%d')
df_fitbit.to_json('./fitbit.json',date_format='iso')
df_fitbit['ActivityDate'] = pd.to_datetime(df_fitbit['ActivityDate'])

In [131]:
fitbit_json = pd.read_json('./fitbit.json')

In [132]:
fitbit_json.head()

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366   2016-03-25       11004           7.11             7.11   
1  1503960366   2016-03-26       17609          11.55            11.55   
2  1503960366   2016-03-27       12736           8.53             8.53   
3  1503960366   2016-03-28       13231           8.93             8.93   
4  1503960366   2016-03-29       12041           7.85             7.85   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.57                      0.46   
1                       0.0                6.92                      0.73   
2                       0.0                4.66                      0.16   
3                       0.0                3.19                      0.79   
4                       0.0                2.16                      1.09   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 4.07                      0.0                 33   
1                 3.91                      0.0                 89   
2                 3.71                      0.0                 56   
3                 4.95                      0.0                 39   
4                 4.61                      0.0                 28   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   12                   205               804      1819  
1                   17                   274               588      2154  
2                    5                   268               605      1944  
3                   20                   224              1080      1932  
4                   28                   243               763      1886

In [133]:
# CONNECTING TO MONGODB ATLAS
try:
    client = pymongo.MongoClient("mongodb+srv://<password>:ineuron@cluster0.ex1ru.mongodb.net/?retryWrites=true&w=majority")
except Exception as e:
    print("Error => {e}")
db = client.test

In [134]:
print(db)

Database(MongoClient(host=['cluster0-shard-00-01.ex1ru.mongodb.net:27017', 'cluster0-shard-00-00.ex1ru.mongodb.net:27017', 'cluster0-shard-00-02.ex1ru.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', authsource='admin', replicaset='atlas-r0vbvb-shard-0', tls=True), 'test')


In [135]:
# CREATING A DATABASE 'fitbit_db' and a new collections 'fitbit' in 'fitbit_db' db
try:
    database = client['fitbit_db']
    collection = database['fitbit']
    collection_attribute = database['fitbit']
except Exception as e:
    print("Error in creating database and collection [dress,sales]")
    print(f"Error => {e}")

In [143]:
fitbit_json.index = fitbit_json.index.map(str)

In [144]:
fitbit_dict =fitbit_json.to_dict('index')

In [145]:
# fitbit_dict

> Insert data into Mongo Db

In [155]:
# collection.insert_one(fitbit_dict)

In [142]:
# fitbit_json.fillna(-1, inplace=True)

In [156]:
# fitbit_json.isna().sum()

# Converting all the dates avaible in dataset to timestamp format in pandas and in sql you to convert it in date format.

In [8]:
df_fitbit['ActivityDate'] = pd.to_datetime(df_fitbit['ActivityDate'])

In [9]:
df_fitbit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 457 entries, 0 to 456
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Id                        457 non-null    int64         
 1   ActivityDate              457 non-null    datetime64[ns]
 2   TotalSteps                457 non-null    int64         
 3   TotalDistance             457 non-null    float64       
 4   TrackerDistance           457 non-null    float64       
 5   LoggedActivitiesDistance  457 non-null    float64       
 6   VeryActiveDistance        457 non-null    float64       
 7   ModeratelyActiveDistance  457 non-null    float64       
 8   LightActiveDistance       457 non-null    float64       
 9   SedentaryActiveDistance   457 non-null    float64       
 10  VeryActiveMinutes         457 non-null    int64         
 11  FairlyActiveMinutes       457 non-null    int64         
 12  LightlyActiveMinutes  

# Find out in this data that how many unique id's we have 

In [10]:
df_fitbit['Id'].nunique()

35

We have a total of 35 unique Id's. 

In [11]:
df_fitbit['Id'].unique()

array([1503960366, 1624580081, 1644430081, 1844505072, 1927972279,
       2022484408, 2026352035, 2320127002, 2347167796, 2873212765,
       2891001357, 3372868164, 3977333714, 4020332650, 4057192912,
       4319703577, 4388161847, 4445114986, 4558609924, 4702921684,
       5553957443, 5577150313, 6117666160, 6290855005, 6391747486,
       6775888955, 6962181067, 7007744171, 7086361926, 8053475328,
       8253242879, 8378563200, 8583815059, 8792009665, 8877689391])

In [12]:
df_fitbit['Id'].value_counts()

4057192912    32
4020332650    32
1503960366    19
1624580081    19
2347167796    15
4702921684    15
4445114986    15
6962181067    14
2320127002    12
4558609924    12
8792009665    12
8378563200    12
8253242879    12
7086361926    12
7007744171    12
2026352035    12
5553957443    12
8877689391    12
1844505072    12
2022484408    12
2873212765    12
3977333714    12
1927972279    12
4319703577    12
8053475328    11
5577150313    11
6290855005    10
1644430081    10
6117666160    10
3372868164    10
6391747486     9
6775888955     9
2891001357     8
8583815059     8
4388161847     8
Name: Id, dtype: int64

# which id is one of the active id that you have in whole dataset 

In [13]:
df_fitbit.sample(6)

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
227  4057192912   2016-04-11        6801           4.98             4.98   
245  4388161847   2016-04-02           0           0.00             0.00   
359  6962181067   2016-04-06       12480           8.86             8.25   
319  6117666160   2016-04-06       12409           9.37             9.37   
130  2873212765   2016-04-08        7157           4.81             4.81   
297  5553957443   2016-04-07       15444          10.08            10.08   

     LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
227                  0.000000                0.00                      0.00   
245                  0.000000                0.00                      0.00   
359                  3.972795                4.18                      0.69   
319                  0.000000                0.00                      0.00   
130                  0.000000                0.00                      0.00   
297                  0.000000                3.27                      1.73   

     LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
227                 4.97                      0.0                  0   
245                 0.00                      0.0                  0   
359                 4.00                      0.0                 55   
319                 9.37                      0.0                  0   
130                 4.81                      0.0                  0   
297                 5.08                      0.0                 51   

     FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
227                    0                   193              1247      2393  
245                    0                     0              1440      1878  
359                   14                   276               593      2170  
319                    0                   491               388      2694  
130                    0                   380              1060      1944  
297                   39                   243               731      2249

In [14]:
df_fitbit.groupby('Id').sum().nlargest(5, columns=['TotalSteps'])

TotalSteps  TotalDistance  TrackerDistance  \
Id                                                       
1503960366      221170     144.540001       144.540001   
8877689391      209005     169.089998       169.089998   
4020332650      184851     132.600000       132.600000   
6962181067      176956     121.070000       117.400001   
8053475328      163288     127.340000       127.340000   

            LoggedActivitiesDistance  VeryActiveDistance  \
Id                                                         
1503960366                  0.000000               52.89   
8877689391                  0.000000               87.39   
4020332650                  0.000000                5.30   
6962181067                 14.052416               36.13   
8053475328                  0.000000               93.93   

            ModeratelyActiveDistance  LightActiveDistance  \
Id                                                          
1503960366                     12.07            79.390000   
8877689391                      5.80            75.270001   
4020332650                      6.29            71.710000   
6962181067                     24.99            59.850000   
8053475328                      5.43            27.950000   

            SedentaryActiveDistance  VeryActiveMinutes  FairlyActiveMinutes  \
Id                                                                            
1503960366                     0.00                681                  301   
8877689391                     0.03                802                  181   
4020332650                     0.20                130                  245   
6962181067                     0.10                495                  406   
8053475328                     0.00                933                  125   

            LightlyActiveMinutes  SedentaryMinutes  Calories  
Id                                                            
1503960366                  4336             15387     34128  
8877689391                  2892             12554     41414  
4020332650                  4172             34663     98412  
6962181067                  3620              8488     29249  
8053475328                  1698             12259     31825

If we assume that the User with most number of steps is the most active person, then User with Id `1503960366` is the most ctive person

# how many of them have not logged there activity find out in terms of number of ids 

In [15]:
df_fitbit.head(3)

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366   2016-03-25       11004           7.11             7.11   
1  1503960366   2016-03-26       17609          11.55            11.55   
2  1503960366   2016-03-27       12736           8.53             8.53   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.57                      0.46   
1                       0.0                6.92                      0.73   
2                       0.0                4.66                      0.16   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 4.07                      0.0                 33   
1                 3.91                      0.0                 89   
2                 3.71                      0.0                 56   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   12                   205               804      1819  
1                   17                   274               588      2154  
2                    5                   268               605      1944

In [16]:
df_fitbit.query('LoggedActivitiesDistance == 0')

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0    1503960366   2016-03-25       11004       7.110000         7.110000   
1    1503960366   2016-03-26       17609      11.550000        11.550000   
2    1503960366   2016-03-27       12736       8.530000         8.530000   
3    1503960366   2016-03-28       13231       8.930000         8.930000   
4    1503960366   2016-03-29       12041       7.850000         7.850000   
..          ...          ...         ...            ...              ...   
452  8877689391   2016-04-08       23014      20.389999        20.389999   
453  8877689391   2016-04-09       16470       8.070000         8.070000   
454  8877689391   2016-04-10       28497      27.530001        27.530001   
455  8877689391   2016-04-11       10622       8.060000         8.060000   
456  8877689391   2016-04-12        2350       1.780000         1.780000   

     LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                         0.0                2.57                      0.46   
1                         0.0                6.92                      0.73   
2                         0.0                4.66                      0.16   
3                         0.0                3.19                      0.79   
4                         0.0                2.16                      1.09   
..                        ...                 ...                       ...   
452                       0.0               11.10                      0.63   
453                       0.0                0.00                      0.02   
454                       0.0               21.92                      1.12   
455                       0.0                1.47                      0.15   
456                       0.0                0.00                      0.00   

     LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                   4.07                     0.00                 33   
1                   3.91                     0.00                 89   
2                   3.71                     0.00                 56   
3                   4.95                     0.00                 39   
4                   4.61                     0.00                 28   
..                   ...                      ...                ...   
452                 8.62                     0.00                 70   
453                 8.02                     0.00                 90   
454                 4.46                     0.00                128   
455                 6.37                     0.01                 18   
456                 1.78                     0.00                  0   

     FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                     12                   205               804      1819  
1                     17                   274               588      2154  
2                      5                   268               605      1944  
3                     20                   224              1080      1932  
4                     28                   243               763      1886  
..                   ...                   ...               ...       ...  
452                   29                   359               982      4196  
453                    9                   289              1052      3841  
454                   46                   211              1055      4526  
455                    7                   225              1190      2820  
456                    0                    58               531       938  

[433 rows x 15 columns]

Out of 457 activities, 433 weren't logged by the users

In [17]:
df_fitbit.groupby('Id')['LoggedActivitiesDistance'].sum().sort_values(ascending=False)

Id
7007744171    34.351326
8378563200    15.127832
6962181067    14.052416
2891001357     9.656064
6391747486     8.754830
4057192912     0.055843
5577150313     0.000000
6117666160     0.000000
6290855005     0.000000
6775888955     0.000000
1503960366     0.000000
4702921684     0.000000
7086361926     0.000000
8053475328     0.000000
8253242879     0.000000
8583815059     0.000000
8792009665     0.000000
5553957443     0.000000
4445114986     0.000000
4558609924     0.000000
1624580081     0.000000
4388161847     0.000000
4319703577     0.000000
4020332650     0.000000
3977333714     0.000000
3372868164     0.000000
2873212765     0.000000
2347167796     0.000000
2320127002     0.000000
2026352035     0.000000
2022484408     0.000000
1927972279     0.000000
1844505072     0.000000
1644430081     0.000000
8877689391     0.000000
Name: LoggedActivitiesDistance, dtype: float64

Out of 35 unique users, only 5 of them have logged their activities. So, a total of 30 Users didn't log their activities

# Find out who is the laziest person id that we have in dataset 

In [18]:
df_fitbit.tail(3)

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
454  8877689391   2016-04-10       28497      27.530001        27.530001   
455  8877689391   2016-04-11       10622       8.060000         8.060000   
456  8877689391   2016-04-12        2350       1.780000         1.780000   

     LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
454                       0.0               21.92                      1.12   
455                       0.0                1.47                      0.15   
456                       0.0                0.00                      0.00   

     LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
454                 4.46                     0.00                128   
455                 6.37                     0.01                 18   
456                 1.78                     0.00                  0   

     FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
454                   46                   211              1055      4526  
455                    7                   225              1190      2820  
456                    0                    58               531       938

### If we consider `Distance` as a criteria for finding Lazy person. Here, we will find percentage of lazy distance a user walks when compared with Total Distance.

> NOTE : Here we are considering (SedentaryActiveDistance+LightActiveDistance) as a LAzy Distance

In [19]:
id_distance = df_fitbit.groupby('Id').aggregate(
    {'LightActiveDistance':'sum','SedentaryActiveDistance':'sum','VeryActiveDistance':'sum','ModeratelyActiveDistance':'sum'}
    ).reset_index()

In [20]:
id_distance.head(8)

Id  LightActiveDistance  SedentaryActiveDistance  \
0  1503960366            79.390000                     0.00   
1  1624580081            50.530000                     0.10   
2  1644430081            35.800000                     0.08   
3  1844505072            27.890000                     0.00   
4  1927972279            17.390000                     0.00   
5  2022484408            57.670000                     0.00   
6  2026352035            25.240000                     0.00   
7  2320127002            23.979999                     0.00   

   VeryActiveDistance  ModeratelyActiveDistance  
0               52.89                     12.07  
1                0.98                      0.52  
2               11.34                     20.30  
3                0.61                      0.37  
4                0.00                      0.74  
5               37.03                     10.61  
6                0.00                      0.00  
7                0.95                      0.54

In [21]:
lazy_distance = (id_distance.iloc[:,1] + id_distance.iloc[:,2])
total_distance = id_distance.iloc[:,1] + id_distance.iloc[:,2] + id_distance.iloc[:,4] + id_distance.iloc[:,4] 

id_distance['lazy_distance_percentage'] = ((lazy_distance/total_distance))*100

In [22]:
id_distance.sample(6)

Id  LightActiveDistance  SedentaryActiveDistance  \
22  6117666160            59.090000                     0.00   
7   2320127002            23.979999                     0.00   
28  7086361926            23.920000                     0.06   
16  4388161847             0.000000                     0.00   
8   2347167796            65.809999                     0.00   
26  6962181067            59.850000                     0.10   

    VeryActiveDistance  ModeratelyActiveDistance  lazy_distance_percentage  
22                0.42                      2.73                 91.541441  
7                 0.95                      0.54                 95.690343  
28               18.35                      6.66                 64.289544  
16                0.00                      0.00                       NaN  
8                12.92                     18.91                 63.504776  
26               36.13                     24.99                 54.534704

In [23]:
id_distance.sort_values(['lazy_distance_percentage'], ascending=False)[:5]

Id  LightActiveDistance  SedentaryActiveDistance  \
6   2026352035                25.24                      0.0   
10  2891001357                 4.83                      0.0   
24  6391747486                 7.59                      0.0   
1   1624580081                50.53                      0.1   
3   1844505072                27.89                      0.0   

    VeryActiveDistance  ModeratelyActiveDistance  lazy_distance_percentage  
6                 0.00                      0.00                100.000000  
10                0.00                      0.00                100.000000  
24                2.03                      0.05                 98.699610  
1                 0.98                      0.52                 97.987227  
3                 0.61                      0.37                 97.415299

-> Users with Id `2026352035` and `2891001357` has highest percentage in terms of Lazy Distance Percentage

### If we consider `TotalSteps` walked by a User as a measuring criteria for Laziness

In [24]:
df_fitbit.head()

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366   2016-03-25       11004           7.11             7.11   
1  1503960366   2016-03-26       17609          11.55            11.55   
2  1503960366   2016-03-27       12736           8.53             8.53   
3  1503960366   2016-03-28       13231           8.93             8.93   
4  1503960366   2016-03-29       12041           7.85             7.85   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.57                      0.46   
1                       0.0                6.92                      0.73   
2                       0.0                4.66                      0.16   
3                       0.0                3.19                      0.79   
4                       0.0                2.16                      1.09   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 4.07                      0.0                 33   
1                 3.91                      0.0                 89   
2                 3.71                      0.0                 56   
3                 4.95                      0.0                 39   
4                 4.61                      0.0                 28   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   12                   205               804      1819  
1                   17                   274               588      2154  
2                    5                   268               605      1944  
3                   20                   224              1080      1932  
4                   28                   243               763      1886

In [25]:
total_steps_per_user = df_fitbit.groupby('Id').aggregate({'TotalSteps':'sum'})
total_steps_per_user.sample(3)

TotalSteps
Id                    
4388161847           0
7086361926       73247
1844505072       43687

In [26]:
total_steps_per_user.nlargest(1,columns=['TotalSteps'])

TotalSteps
Id                    
1503960366      221170

In [27]:
total_steps_per_user.sort_values(by=['TotalSteps'], ascending=True)[:4]

TotalSteps
Id                    
4388161847           0
2891001357        6189
6391747486       12032
6290855005       16181

-> User with id `4388161847` walked 0 steps(Lowest) followed by user with Id `2891001357` with 6189 steps

# Explore over an internet that how much calories burn is required for a healthy person and find out how many healthy person we have in our dataset 

In [28]:
df_fitbit.sample(5)

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
204  4057192912   2016-03-19           0           0.00             0.00   
220  4057192912   2016-04-04           0           0.00             0.00   
375  7007744171   2016-04-10       10146           6.79             6.79   
433  8792009665   2016-04-01        4592           2.94             2.94   
153  3977333714   2016-04-01        4499           3.01             3.01   

     LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
204                       0.0                0.00                      0.00   
220                       0.0                0.00                      0.00   
375                       0.0                0.07                      0.61   
433                       0.0                0.10                      0.14   
153                       0.0                0.00                      0.00   

     LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
204                 0.00                      0.0                  0   
220                 0.00                      0.0                  0   
375                 6.11                      0.0                  1   
433                 2.70                      0.0                  4   
153                 3.01                      0.0                  0   

     FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
204                    0                     0              1440      1777  
220                    0                     0              1440      1776  
375                   12                   397              1030      2700  
433                    8                   176               748      2260  
153                    0                   168               842      1288

In [29]:
df_fitbit['ActivityDate']

0     2016-03-25
1     2016-03-26
2     2016-03-27
3     2016-03-28
4     2016-03-29
         ...    
452   2016-04-08
453   2016-04-09
454   2016-04-10
455   2016-04-11
456   2016-04-12
Name: ActivityDate, Length: 457, dtype: datetime64[ns]

In [30]:
calories_per_month_User = df_fitbit.groupby(['Id',df_fitbit['ActivityDate'].dt.month]).aggregate({'Calories':'sum'}).reset_index()
calories_per_month_User.head()
# NOTE : Here, ActivityDate indicates Month

Id  ActivityDate  Calories
0  1503960366             3     13444
1  1503960366             4     20684
2  1624580081             3      9440
3  1624580081             4     16265
4  1644430081             4     29164

> A general rule is to aim to burn 400 to 500 calories, five days a week during your workouts.

> We have on an average 4.3 weeks in a month. So, Healthy person should bur around (4*450)+(450) = 

# how many person are not a regular person with respect to activity try to find out those 

In [31]:
df_fitbit.head()

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366   2016-03-25       11004           7.11             7.11   
1  1503960366   2016-03-26       17609          11.55            11.55   
2  1503960366   2016-03-27       12736           8.53             8.53   
3  1503960366   2016-03-28       13231           8.93             8.93   
4  1503960366   2016-03-29       12041           7.85             7.85   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.57                      0.46   
1                       0.0                6.92                      0.73   
2                       0.0                4.66                      0.16   
3                       0.0                3.19                      0.79   
4                       0.0                2.16                      1.09   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 4.07                      0.0                 33   
1                 3.91                      0.0                 89   
2                 3.71                      0.0                 56   
3                 4.95                      0.0                 39   
4                 4.61                      0.0                 28   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   12                   205               804      1819  
1                   17                   274               588      2154  
2                    5                   268               605      1944  
3                   20                   224              1080      1932  
4                   28                   243               763      1886

In [32]:
df_fitbit['ActivityDay'] = df_fitbit['ActivityDate'].dt.day
df_fitbit['ActivityMonth'] = df_fitbit['ActivityDate'].dt.month

In [33]:
df_fitbit.sample(3)

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
110  2347167796   2016-03-31        6760           4.47             4.47   
17   1503960366   2016-04-11       10990           7.26             7.26   
244  4388161847   2016-04-01           0           0.00             0.00   

     LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
110                       0.0                0.00                      0.00   
17                        0.0                2.04                      0.57   
244                       0.0                0.00                      0.00   

     LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
110                 4.47                      0.0                  0   
17                  4.65                      0.0                 26   
244                 0.00                      0.0                  0   

     FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  \
110                    0                   250               613      1929   
17                    14                   216               855      1811   
244                    0                     0              1440      1878   

     ActivityDay  ActivityMonth  
110           31              3  
17            11              4  
244            1              4

In [34]:
activities_per_month_user =  df_fitbit.groupby(['Id','ActivityMonth']).aggregate({'ActivityDay':'count'}).reset_index()
activities_per_month_user

Id  ActivityMonth  ActivityDay
0   1503960366              3            7
1   1503960366              4           12
2   1624580081              3            7
3   1624580081              4           12
4   1644430081              4           10
5   1844505072              4           12
6   1927972279              4           12
7   2022484408              4           12
8   2026352035              4           12
9   2320127002              4           12
10  2347167796              3            3
11  2347167796              4           12
12  2873212765              4           12
13  2891001357              3            3
14  2891001357              4            5
15  3372868164              4           10
16  3977333714              4           12
17  4020332650              3           20
18  4020332650              4           12
19  4057192912              3           20
20  4057192912              4           12
21  4319703577              3            3
22  4319703577              4            9
23  4388161847              3            3
24  4388161847              4            5
25  4445114986              3            3
26  4445114986              4           12
27  4558609924              4           12
28  4702921684              3            3
29  4702921684              4           12
30  5553957443              4           12
31  5577150313              4           11
32  6117666160              4           10
33  6290855005              4           10
34  6391747486              4            9
35  6775888955              4            9
36  6962181067              3            2
37  6962181067              4           12
38  7007744171              4           12
39  7086361926              4           12
40  8053475328              4           11
41  8253242879              4           12
42  8378563200              4           12
43  8583815059              4            8
44  8792009665              4           12
45  8877689391              4           12

> Here we are considering 12 as a threshold number(days) per month to classify between REGULAR and IRREGULAR

In [35]:
threshold_days = 10
activities_per_month_user.query('ActivityDay < @threshold_days')['Id'].unique().size

12

In [36]:
activities_per_month_user.query('ActivityDay < @threshold_days')['Id'].unique()

array([1503960366, 1624580081, 2347167796, 2891001357, 4319703577,
       4388161847, 4445114986, 4702921684, 6391747486, 6775888955,
       6962181067, 8583815059])

- Total of 12 persons are irregular. Their Id's are [1503960366, 1624580081, 2347167796, 2891001357, 4319703577,4388161847, 4445114986, 4702921684, 6391747486, 6775888955,6962181067, 8583815059]

# who is the thired most active person in this dataset find out those in pandas and in sql both 

> Here   `lazy_distance_percentage` represents No. of lazy steps(SedentaryActiveDistance+LightActiveDistance) a User has walked when compared to Total Steps walked in terms of percentage. So, **Higher** the percentage, Lazy the person is and Vice-Versa

In [37]:
id_distance.sample(4)

Id  LightActiveDistance  SedentaryActiveDistance  \
11  3372868164                33.46                     0.14   
32  8583815059                17.88                     0.00   
30  8253242879                 9.51                     0.03   
0   1503960366                79.39                     0.00   

    VeryActiveDistance  ModeratelyActiveDistance  lazy_distance_percentage  
11                7.44                      1.14                 93.645485  
32                0.37                      0.73                 92.450879  
30                7.24                      3.39                 58.455882  
0                52.89                     12.07                 76.683087

In [38]:
id_distance.sort_values(by=['lazy_distance_percentage']).nsmallest(3,columns=['lazy_distance_percentage']).iloc[2:,:]

Id  LightActiveDistance  SedentaryActiveDistance  \
23  6290855005                 6.53                      0.0   

    VeryActiveDistance  ModeratelyActiveDistance  lazy_distance_percentage  
23                2.02                      3.68                 47.012239

Third most active person is with Id `6290855005`

# who is the 5th most laziest person avilable in dataset find it out 

In [39]:
id_distance.sort_values(by=['lazy_distance_percentage'], ascending=False).iloc[:5,:]

Id  LightActiveDistance  SedentaryActiveDistance  \
6   2026352035                25.24                      0.0   
10  2891001357                 4.83                      0.0   
24  6391747486                 7.59                      0.0   
1   1624580081                50.53                      0.1   
3   1844505072                27.89                      0.0   

    VeryActiveDistance  ModeratelyActiveDistance  lazy_distance_percentage  
6                 0.00                      0.00                100.000000  
10                0.00                      0.00                100.000000  
24                2.03                      0.05                 98.699610  
1                 0.98                      0.52                 97.987227  
3                 0.61                      0.37                 97.415299

5th most Laziest Person is with Id `1844505072`

# what is a totla acumulative calories burn for a person find out 

In [40]:
calories_per_month_User.head()

Id  ActivityDate  Calories
0  1503960366             3     13444
1  1503960366             4     20684
2  1624580081             3      9440
3  1624580081             4     16265
4  1644430081             4     29164

In [41]:
calories_burned_per_user = calories_per_month_User.groupby('Id').aggregate({'Calories':'sum'})

In [42]:
calories_burned_per_user

Calories
Id                  
1503960366     34128
1624580081     25705
1644430081     29164
1844505072     19391
1927972279     27048
2022484408     29704
2026352035     16266
2320127002     18385
2347167796     30320
2873212765     20349
2891001357     18187
3372868164     18603
3977333714     16777
4020332650     98412
4057192912     60913
4319703577     23931
4388161847     14442
4445114986     31617
4558609924     21962
4702921684     42322
5553957443     21630
5577150313     36303
6117666160     20993
6290855005     21656
6391747486     15868
6775888955     24523
6962181067     29249
7007744171     31525
7086361926     26124
8053475328     31825
8253242879     17556
8378563200     40274
8583815059     19130
8792009665     24884
8877689391     41414

In [161]:
from IPython.display import Image

Image(url="./Screenshot 2022-08-12 at 9.23.35 PM.png", width=1000, height=500)

## MYSQL Screenshot

In [162]:
Image(url="./fitbit_sql_ss.png", width=1000, height=500)